pdf.go

```diff
@@ -859,6 +859,54 @@ func (w *pdfPageWriter) EndTextObject() {
 	w.inTextObject = false
 }
 
+// newStringEscaper adds a backlash byte before following bytes: ( ) \
+//
+// According to the PDF spec:
+// When a string is written by enclosing the data in parentheses, bytes whose values are the
+// same as thoseof  the  ASCII  characters  left  parenthesis  (40),  right  parenthesis  (41),
+// and  back-slash  (92)  must  be  preceded  by  a  backslash  character. (...)
+// These  rules  apply  to  each individual  byte  in  a  string  object, whether  the  string
+// is  interpreted  by  the  text-showing operators as single-byte or multiple-byte character codes.
+func newStringEscaper(w io.Writer) escaper {
+	return escaper{
+		w:             w,
+		bytesToEscape: []byte{'\\', '(', ')'},
+		escapeBytes:   []byte{'\\'},
+	}
+}
+
+type escaper struct {
+	w             io.Writer
+	bytesToEscape []byte
+	escapeBytes   []byte
+}
+
+func (r escaper) Write(buf []byte) (n int, err error) {
+	var i, k int
+	for j, b := range buf {
+		if bytes.IndexByte(r.bytesToEscape, b) == -1 {
```

Comment:
Did you choose `IndexByte` because it has ASM implementations (and thus is probably faster)?

Reply:
It is more because it as the only function which directly worked on byte slice & byte :-)

Reply:
Just in case, but b is a rune and not a byte slice. Being utf8, the only possible position for ASCII characters is at the beginning, see https://research.swtch.com/utf8

Reply:
> b is a rune and not a byte slice

In this code, there are only bytes and bytes slices (`b` would have been a rune if `buf` was a string).

> Being utf8...

Nothing guarantees that string is encoded in utf8

```
binary.Write(buf, binary.BigEndian, indices)
```

The content of `indices` is a gathering of values coming from `f.sfnt.GlyphIndex` (which are arbitrary numbers: it has nothing to to with utf8).
`indices` could ends up being: `[0b11000000, 0b000000000]` (which is invalid utf8 according to https://research.swtch.com/utf8, but a valid string)

Quoting https://blog.golang.org/strings
> Some people think Go strings are always UTF-8, but they are not: only string literals are UTF-8. As we showed in the previous section, string values can contain arbitrary bytes; as we showed in this one, string literals always contain UTF-8 text as long as they have no byte-level escapes.

Reply:
Your code still works fine, because (luckily) all characters that are escaped are represented in UTF-8 using 1 byte (and 1-byte UTF-8 chars can't appear in any other valid UTF-8 character).

If an "impossible" UTF-8 byte is met by go (as my example), it will emit a rune containing all bytes and start from scratch using the "impossible" byte (which might well be possible as first byte of a UTF-8 representation).

https://play.golang.org/p/4N2N0r5nfjv

Reply:
Sorry not sure what I was thinking

Reply:
No problem, I also learned a lot along the way :)

_When do you think you would have time to review #25 ?_
_(I don't want to harass you, just want to have an idea, when I can hack further on this project :)_

Reply:
Sorry I'm on a tough deadline regarding covid-19, hopefully I can get back to from monday onwards